btree/btree.c

```diff
@@ -5814,6 +6237,260 @@ static bool vkvv_isfit(struct slot *slot)
+	int                  total_avail_ksize = 0;
+	int                  total_avail_vsize = 0;
+	int                  incoming_ksize    =
+				m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	int                  incoming_vsize    =
+				m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	int                  i;
+
+	if (!IS_EMBEDDED_INDIRECT(slot->s_node))
+		return false;
+
+	if (vkvv_level(slot->s_node) == 0 &&
+	    vkvv_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
+		incoming_vsize += CRC_VALUE_SIZE;
+
+	for (i = 0; i < h->vkvv_dir_entries; i++) {
+		total_avail_ksize += (dir[i].alloc_key_size - dir[i].key_size);
+		total_avail_vsize += (dir[i].alloc_val_size - dir[i].val_size);
+	}
+
+	if ((total_avail_ksize + total_avail_vsize) >= (incoming_ksize +
+	     incoming_vsize + sizeof(*dir)))
+		return true;
+	else
+		return false;
+}
+
+static void vkvv_sort_dir(struct vkvv_dir_rec *dir, int count)
+{
+	int                 i;
+	int                 j;
+	struct vkvv_dir_rec item;
+
+	for (i = 1; i < count; i++) {
+		item = dir[i];
+		j = i - 1;
+		while (j >= 0 && dir[j].key_offset > item.key_offset) {
+			dir[j+1] = dir[j];
+			j = j -1;
+		}
+		dir[j+1] = item;
+	}
+}
+
+static bool vkvv_is_rec_with_hole(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size != 0 && entry.val_size != 0 &&
+	   ((entry.alloc_key_size - entry.key_size > 0) ||
+	    (entry.alloc_val_size - entry.val_size > 0)))
+		return true;
+	else
+		return false;
+}
+
+static bool vkvv_is_rec_without_hole(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size != 0 && entry.val_size != 0 &&
+	   entry.alloc_key_size - entry.key_size == 0 &&
+	   entry.alloc_val_size - entry.val_size == 0)
+		return true;
+	else
+		return false;
+}
+
+static bool vkvv_is_rec_free_frag(struct vkvv_dir_rec entry)
+{
+	if(entry.key_size == 0 && entry.val_size == 0) {
+		M0_ASSERT(entry.alloc_key_size >= 0 ||
+			  entry.alloc_val_size >= 0);
+		return true;
+	}
+	else
+		return false;
+}
+
+static int vkvv_get_idx_of_rec(const struct nd *node, struct vkvv_dir_rec entry)
+{
+	struct vkvv_head    *h   = vkvv_data(node);
+	struct vkvv_dir_rec *dir = vkvv_dir_get(node);
+	int                  i;
+
+	for(i = 0; i < h->vkvv_dir_entries; i++) {
+		if (dir[i].key_offset == entry.key_offset) {
+			M0_ASSERT(dir[i].val_offset == entry.val_offset);
+			break;
+		}
+	}
+	return i;
+}
+
+static void vkvv_compaction(struct slot *slot)
+{
+	struct vkvv_head    *h             = vkvv_data(slot->s_node);
+	struct vkvv_dir_rec *dir           = vkvv_dir_get(slot->s_node);
+	int                  dir_rec_count = h->vkvv_dir_entries;
+	struct vkvv_dir_rec  new_dir[dir_rec_count];
+	struct vkvv_dir_rec  old_dir[dir_rec_count];
+	int                  i             = 0;
+	int                  c1            = 0;
+	int                  c2            = 0;
+
+	memcpy(old_dir, dir, sizeof(*dir) * dir_rec_count);
+	memcpy(new_dir, dir, sizeof(*dir) * dir_rec_count);
+	vkvv_sort_dir(new_dir, dir_rec_count);
```

Comment:
I think, we do not require to copy entire directory. We can copy and sort dir[0] to dir[vkvv_used] as we just want to note the directory entries with valid keys and values.

The code can be modified as follows:
while ( i < vkvv_used):
{
first_ent = new_dir[i]
next_ent = new_dir[i + 1]

curr_khole = next_ent.key_offset - (first_ent.key_offset + first_ent.key_size)
// compact key region these two entries
m0_memmove((void *)h + next_ent.key_offset -
					   curr_khole,
					   (void *)h + next_ent.key_offset,
					   next_ent.alloc_key_size);

// update these directory entries in original directory i.e. dir.
next_ent.key_offset          -= curr_khole;
				new_dir[i+1].key_offset      -= curr_khole;
				dir[next_idx].key_offset     -= curr_khole;
				next_ent.alloc_key_size      += curr_khole;
				new_dir[i+1].alloc_key_size  += curr_khole;
				dir[next_idx].alloc_key_size += curr_khole;
// similarly for value region.
}


In this way we will not require vkvv_is_rec_with_hole, vkvv_is_rec_with_hole, vkvv_is_rec_free_frag

Reply:
Let me know what do you think. It is just a suggestion and can be done later as current code is also fine but lengthy.

Reply:
Lets think over this change after the current code is working fine.